go/vt/vtgate/query_executor.go

```diff
@@ -73,3 +78,105 @@ func (vc *queryExecutor) GetShardForKeyspaceID(allShards []*topodatapb.ShardRefe
 func (vc *queryExecutor) ExecuteShard(keyspace string, shardQueries map[string]querytypes.BoundQuery) (*sqltypes.Result, error) {
 	return vc.router.scatterConn.ExecuteMultiShard(vc.ctx, keyspace, shardQueries, vc.tabletType, NewSafeSession(nil), false, vc.options)
 }
+
+func (vc *queryExecutor) ExecuteShow(query string, bindvars map[string]interface{}, keyspace string) (*sqltypes.Result, error) {
+	if query == sqlparser.ShowDatabasesStr {
+		keyspaces, err := getAllKeyspaces(vc.ctx, vc.router.serv, vc.router.cell)
+		if err != nil {
+			return nil, err
+		}
+
+		rows := make([][]sqltypes.Value, len(keyspaces))
+		for i, v := range keyspaces {
+			rows[i] = []sqltypes.Value{sqltypes.MakeString([]byte(v))}
+		}
+
+		result := &sqltypes.Result{
+			Fields: []*querypb.Field{{
+				Name: "Databases",
+				Type: sqltypes.VarChar,
+			}},
+			RowsAffected: uint64(len(keyspaces)),
+			InsertID:     0,
+			Rows:         rows,
+		}
+
+		return result, nil
+	}
+
+	if query == sqlparser.ShowShardsStr {
+		keyspaces, err := getAllKeyspaces(vc.ctx, vc.router.serv, vc.router.cell)
+		if err != nil {
+			return nil, err
+		}
+
+		var shards []string
+
+		for _, keyspace := range keyspaces {
+			_, _, ks_shards, err := getKeyspaceShards(vc.ctx, vc.router.serv, vc.router.cell, keyspace, vc.tabletType)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, shard := range ks_shards {
+				shards = append(shards, keyspace+"/"+shard.Name)
+			}
+		}
+
+		rows := make([][]sqltypes.Value, len(shards))
+		for i, v := range shards {
+			rows[i] = []sqltypes.Value{sqltypes.MakeString([]byte(v))}
+		}
+
+		result := &sqltypes.Result{
+			Fields: []*querypb.Field{{
+				Name: "Shards",
+				Type: sqltypes.VarChar,
+			}},
+			RowsAffected: uint64(len(shards)),
+			InsertID:     0,
+			Rows:         rows,
+		}
+
+		return result, nil
+	}
+
+	if query == sqlparser.ShowTablesStr {
```

Comment:
TIL - I didn't think that `show tables` worked at all, but you're right that especially with the upcoming work to target a query to a shard, this is something we should support.

So I'll go with `show vschema_tables` for this case.

One question though, in a sharded keyspace, what should `show tables` to actually do? I presume it could pass through to the first shard?